libs/src/services/solanaWeb3Manager/rewards.js

```diff
@@ -0,0 +1,767 @@
+      isSigner: false,
+      isWritable: false,
+    },
+    {
+      pubkey: SystemProgram.programId,
+      isSigner: false,
+      isWritable: false,
+    },
+  ]
+
+  const instructionData = new SubmitAttestationInstructionData({ transferId })
+  const serializedInstructionData = borsh.serialize(
+    submitAttestationInstructionSchema,
+    instructionData
+  )
+  const serializedInstructionEnum = Buffer.from(
+    Uint8Array.of(SUBMIT_INSTRUCTION_ENUM_VALUE, ...serializedInstructionData)
+  )
+
+  return new TransactionInstruction({
+    keys: verifyInstructionAccounts,
+    programId: rewardManagerProgramId,
+    data: serializedInstructionEnum,
+  })
+}
+
+/**
+ *
+ * @param {{
+ *   attestationMeta: AttestationMeta
+ *   isOracle: boolean
+ *   recipientEthAddress: string
+ *   tokenAmount: BN
+ *   transferId: string
+ *   oracleAddress: string
+ *   instructionIndex: number
+ * }} {
+ *   attestationMeta,
+ *   isOracle,
+ *   recipientEthAddress,
+ *   tokenAmount,
+ *   transferId,
+ *   oracleAddress,
+ *   instructionIndex
+ * }
+ * @returns {TransactionInstruction}
+ */
+const generateSecpInstruction = ({
+  attestationMeta,
+  isOracle,
+  recipientEthAddress,
+  tokenAmount,
+  transferId,
+  oracleAddress,
+  instructionIndex,
+}) => {
+  // Perform signature manipulations:
+  // - remove the 0x prefix
+  // - lose the final byte / recovery ID
+  let strippedSignature = attestationMeta.signature.replace("0x", "")
+  const recoveryIdStr = strippedSignature.slice(strippedSignature.length - 2)
+  const recoveryId = new BN(recoveryIdStr, "hex").toNumber()
+  strippedSignature = strippedSignature.slice(0, strippedSignature.length - 2)
+  const encodedSignature = Uint8Array.of(
+    ...new BN(strippedSignature, "hex").toArray("be")
+  )
+
+  const encodedSenderMessage = constructAttestation(
+    isOracle,
+    recipientEthAddress,
+    tokenAmount,
+    transferId,
+    oracleAddress
+  )
+
+  return Secp256k1Program.createInstructionWithEthAddress({
+    ethAddress: attestationMeta.ethAddress,
+    message: encodedSenderMessage,
+    signature: encodedSignature,
+    recoveryId,
+    instructionIndex,
+  })
+}
+
+// Misc
+
+/**
+ * Converts an eth address hex represenatation to an array of Uint8s in big endian notation
+ * @param {string} ethAddress
+ * @returns {Uint8Array}
+ */
+const ethAddressToArray = (ethAddress) => {
+  const strippedEthAddress = ethAddress.replace("0x", "")
+  return Uint8Array.of(...new BN(strippedEthAddress, "hex").toArray("be"))
+}
+
+/**
+ * Constructs a transfer ID
+ * @param {string} challengeId
+ * @param {string} specifier
+ * @returns {string}
+ */
+const constructTransferId = (challengeId, specifier) =>
+  `${challengeId}:${specifier}`
+
+/**
+ * Converts a BN to a Uint8Array of length 8, in little endian notation.
+ * Useful for when Rust wants a u64 (8 * 8) represented as a byte array.
+ *
+ * @param {BN} bn
+ */
+const padBNToUint8Array = (bn) => bn.toArray("le", 8)
```

Comment:
Not sure I totally follow why this needs to be little endian specifically.
isnt this kinda done on the program side?
https://github.com/AudiusProject/audius-protocol/blob/master/solana-programs/reward-manager/program/src/processor.rs#L389

(maybe just a code pointer here would be helpful)

Reply:
It needs to be LE on the client because we construct the attestation here in the client, and then the Solana program recreates it later in LE and compares it to what the client sent it. Can put a codepointer